### PR TITLE
[TEP-0050] Implement PipelineTask OnError

### PR DESCRIPTION
**Pull request by QuanZhang-William**
# Changes

Part of [#7165][#7165]. In [TEP-0050][tep-0050], we proposed to add an `OnError` API field under `PipelineTask` to configure error handling strategy.

This commits leverages `PipelineTask.OnError` API field introduced in the previous PR, implement the error handling strategy, update related docs and tests.

/kind feature

[tep-0050]: https://github.com/tektoncd/community/blob/main/teps/0050-ignore-task-failures.md
[#7165]: https://github.com/tektoncd/pipeline/issues/7165

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [ ] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Implement "Ignore Task Failure" with new "PipelineTask.OnError" API field (TEP-0050). User can now set `pipelineTask.onError: continue` to ignore failure
```

### CODE REVIEW

File: pkg/apis/pipeline/v1/pipelinerun_types.go
Comment (dibyom) [inline pkg/apis/pipeline/v1/pipelinerun_types.go:415]: nit: probably worth mentioning here that the failure strategy is from the pipelineTask's onError field

Comment (QuanZhang-William) [reply pkg/apis/pipeline/v1/pipelinerun_types.go]: SG! I have updated the commnet, PTAL

File: pkg/pod/status.go
Comment (QuanZhang-William) [reply pkg/pod/status.go:133]: Hi @JeromeJu. This is a good questions.

TEP-0050 is designed with requirement that `the task would be considered "successful" for the purposes of determining the status of the pipeline` (i.e. the `TaskRun` itself should still be failed, but with reason `FailureIgnored`). Looking at the [usecases](https://github.com/tektoncd/community/blob/main/teps/0050-ignore-task-failures.md#use-cases) in TEP-0050, it is mainly used to ignore pod error itself. 

I think your question is that for the validation/reconciler errors (like the `pvc creation`), should we also mark the reason as `FailureIgnored`? 

I think it might **not** be a good idea make the validation/reconciler errors  with reason `FailureIgnored`:
  - The validation/reconciler errors is likely from the Pipeline/Task Author, it shouldn't be ignored, but should be raised and fixed, `FailureIgnored` error reason could be confusing.
  - the error reason `FailureIgnored` becomes too generic if we use it across the TaskRun reconciler code. Instead, I'd prefer to keep the scope of `FailureIgnored` reason to the pod execution time.

WDYT?

**Comment by QuanZhang-William** [comment]
cc @tektoncd/core-collaborators @tektoncd/core-maintainers @jerop @pritidesai . Please take a look if you are interested!

File: pkg/pod/status.go
Comment (JeromeJu) [reply pkg/pod/status.go:133]: Thanks @QuanZhang-William for the offline conversation and completing the 2nd half of my question!

And yes, I do agree that in actual use cases, only the container(Step) failures in the pod, which indicates users' own test failures, shall be ignored but not the "preparation" of those steps i.e. pvc creation and validations. Thanks for the explanation here.

So I think the current implementation is the way to go. One small ask is maybe we would can add this explanation/conversation somewhere in the docString or maybe link to the TEP050 for this design choice.

Comment (dibyom) [reply pkg/pod/status.go:133]: This makes sense as long as we document that `FailedIgnored` is ignoring a failure when the actual task execution fails but not otherwise (validation failure, etc.)

**Comment by dibyom** [comment]
Overall, this looks good to me! Thanks @QuanZhang-William Once you address the remaining open comments, I can approve.

File: test/ignore_task_error_test.go
Comment (QuanZhang-William) [reply test/ignore_task_error_test.go]: Fixed, thanks!

File: pkg/pod/status.go
Comment (QuanZhang-William) [reply pkg/pod/status.go:133]: SG. I have added the docstring. PTAL

**Comment by QuanZhang-William** [comment]
> Overall, this looks good to me! Thanks @QuanZhang-William Once you address the remaining open comments, I can approve.

Thanks @dibyom. I've addressed the comments. PTAL!

File: pkg/reconciler/pipelinerun/resources/pipelinerunstate.go
Comment (JeromeJu) [inline pkg/reconciler/pipelinerun/resources/pipelinerunstate.go:283]: nit: "stopAndFail" is the default here this sounds like users have to set OnError?
Will leave this up to your decision.

**Review by JeromeJu — APPROVED** [review]
Thanks @QuanZhang-William 
One nit on release note, might be helpful to also include how users could use with the OnError feature i.e. "users can now set OnError to 'continue' to ignore failure from previous task".

**Review by dibyom — APPROVED** [review]
Any other feedback on this before merge? cc @vdemeester @jerop @pritidesai @afrittoli

**Comment by QuanZhang-William** [comment]
> Any other feedback on this before merge? cc @vdemeester @jerop @pritidesai @afrittoli

Hi @dibyom. This PR has been opened for more than a week. Was wondering if we can merge? 🙏

**Review by Yongxuanzhang — COMMENTED** [review]
/lgtm

**Pull request merged**
